### PR TITLE
Overhauling the Metrics

### PR DESCRIPTION
**Pull request by rbharath**
`dc.metrics` is the buggiest and least ergonomic part of DeepChem. This is due primarily to the inconsistent handling of shapes. This WIP PR is working towards cleaning up the metrics handling

### CODE REVIEW

**Comment by rbharath** [comment]
I think this PR is ready for a first round of review. It makes the following changes:

- Cleans up shape handling in metrics.
- Adds lots of documentation
- Allows `Evaluator` and `model.evaluate` to accept sklearn metrics directly rather than having to wrap in `dc.metrics.Metric` objects.
- Fixes multiclass metric handling.

@peastman @nd-02110114 Could you take a first review pass at this PR?

File: deepchem/metrics/__init__.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import warnings
 import sklearn.metrics
 import logging
+# TODO: Imported metrics will be removed in a futrue version of DeepCHem
```
Comment (peastman) [inline deepchem/metrics/__init__.py:7]: Do we really want to do that?  I don't think it does any harm to import them here, and it's easier to be able to get all metrics from `dc.metrics` instead of having to remember which ones are there and which ones are in sklearn.

File: deepchem/metrics/__init__.py [hunk 2]
```diff
@@ -19,11 +20,280 @@
...
+
+  Parameters
+  ----------
+  y: np.ndarray
+    Must have shape `(N, n_classes)` and be class probabilities.
+  threshold: float, optional (Default 0.5)
+    The threshold probability for the positive class. Note that this
+    threshold will only be applied for binary classifiers (where
+    `n_classes==2`). If specified for multiclass problems, will be
+    ignored.
```
Comment (peastman) [inline deepchem/metrics/__init__.py:104]: This creates opportunities for bugs, where users don't realize the argument is being ignored.  Instead how about making the default be `None`, and throwing an exception if they set it to anything else for a multiclass problem?

File: deepchem/metrics/__init__.py [hunk 3]
```diff
@@ -314,132 +609,131 @@ def __init__(self,
...
+      must be of shape `(N, n_tasks)` if a regression metric.
     y_pred: np.ndarray
-      An np.ndarray containing predicted values for each task.
+      An np.ndarray containing predicted values for each task. Must be
+      of shape `(N, n_tasks, n_classes)` if a classification metric,
+      else must be of shape `(N, n_tasks)` if a regression metric.
     w: np.ndarray, optional
-      An np.ndarray containing weights for each datapoint.
+      An np.ndarray containing weights for each datapoint. If
+      specified,  must be of shape `(N, n_tasks)`.
```
Comment (peastman) [inline deepchem/metrics/__init__.py:640]: Except they don't actually have to have those shapes, because the methods you wrote make heroic efforts to figure out how to interpret things of all sorts of different shapes!

File: deepchem/metrics/tests/test_normalize.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                                                                          0]])
+    yp = from_one_hot(y_hot)
+    assert np.array_equal(expected, y_hot)
+    assert np.array_equal(y, yp)
+
+  def test_normalize_scalar_classification_binary(self):
+    """Tests 1d classification normalization."""
+    y = 1 
+    y_out = normalize_prediction_shape(y, mode="classification")
+    assert y_out.shape == (1, 1, 2)
```
Comment (peastman) [inline deepchem/metrics/tests/test_normalize.py:30]: In all the tests in this file, we should check that the output actually has the correct value, not just the correct shape.

File: deepchem/models/sklearn_models/__init__.py [hunk 5]
```diff
@@ -21,29 +22,34 @@
...
-  """
-  Abstract base class for different ML models.
+  """Wrapper class that wraps scikit-learn models as DeepChem models.
+
+  When you're working with scikit-learn and DeepChem, at times it can
+  be useful to wrap a scikit-learn model as a DeepChem model. The
+  reason for this might be that you want to do an apples-to-apples
+  comparison of a scikit-learn model to another DeepChem model, or
+  perhaps you want to use the hyperparameter tuning capabilities in
+  `dc.hyper`. The `SklearnModel` class provides a
```
Comment (peastman) [inline deepchem/models/sklearn_models/__init__.py:36]: What does it provide?

File: deepchem/utils/evaluate.py [hunk 6]
```diff
@@ -2,122 +2,319 @@
...
+  Parameters
+  ----------
+  dataset: dc.data.Dataset
+    Dataset on which predictions have been made.
+  y_preds: np.ndarray
+    Predictions to output
+  csv_out: str
+    Name of file to write predictions to.
+  """
+  mol_ids = dataset.ids
```
Comment (nd-02110114) [inline deepchem/utils/evaluate.py:55]: Currently, we start supporting inorganic crystals. So, I think this variable name is not good.

File: deepchem/utils/evaluate.py [hunk 7]
```diff
@@ -2,122 +2,319 @@
...
+
+  Parameters
+  ----------
+  dataset: dc.data.Dataset
+    Dataset on which predictions have been made.
+  y_preds: np.ndarray
+    Predictions to output
+  csv_out: str
+    Name of file to write predictions to.
+  """
```
Comment (nd-02110114) [inline deepchem/utils/evaluate.py:54]: Is the depricated sign needed like `output_statistics `?
```
logger.warning("output_statistics is deprecated.")
```

File: deepchem/metrics/__init__.py [hunk 8]
```diff
@@ -19,11 +20,280 @@
...
+    `n_classes=2` for the case of scalars. Note this parameter only
+    has value if `mode=="classification"`
+
+  Returns
+  -------
+  y_out: np.ndarray
+    If `mode=="classification"`, `y_out` is an array of shape `(N,
+    n_tasks, n_classes)`. If `mode=="regression"`, `y_out` is an array
+    of shape `(N, n_tasks)`.
+  """
```
[the remark this answers is not in the document]
Comment (nd-02110114) [reply deepchem/metrics/__init__.py:224]: I try to refactor! Please refer to my codes https://github.com/deepchem/deepchem/pull/1937

File: deepchem/metrics/__init__.py [hunk 9]
```diff
@@ -270,38 +543,60 @@ def __init__(self,
...
+      )
+    else:
+      self.compute_energy_metric = False
     self.metric = metric
-    self.task_averager = task_averager
-    self.is_multitask = (self.task_averager is not None)
+    if task_averager is None:
+      self.task_averager = np.mean
+    else:
+      self.task_averager = task_averager
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/metrics/__init__.py:584]: That doesn't do the same thing.  Lots of values other than None can evaluate to False, so it's always better to check for None explicitly.

File: deepchem/metrics/__init__.py (hunk 1 above)
Comment (rbharath) [reply deepchem/metrics/__init__.py:7]: Good point. I think you're right. I'll remove these deprecations and add them to the docs.

File: deepchem/metrics/tests/test_normalize.py (hunk 4 above)
Comment (rbharath) [reply deepchem/metrics/tests/test_normalize.py:30]: Good suggestion! Adding in tests for correct values

File: deepchem/models/sklearn_models/__init__.py (hunk 5 above)
Comment (rbharath) [reply deepchem/models/sklearn_models/__init__.py]: Ah forgot to finish this description. Will flesh out this docstring!

File: deepchem/metrics/__init__.py [hunk 10]
```diff
@@ -19,11 +20,280 @@
...
+  >>> y = np.random.rand(10)
+  >>> y_out = normalize_prediction_shape(y, "regression")
+  >>> y_out.shape
+  (10, 1)
+
+  Parameters
+  ----------
+  y: np.ndarray
+    If `mode=="classification"`, `y` is an array of shape `(N,)` or
+    `(N, n_classes)` or `(N, n_tasks, n_classes)`. If `y` is of shape
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py]: Will fix!

File: deepchem/metrics/__init__.py [hunk 11]
```diff
@@ -19,11 +20,280 @@
...
+
+  This utility function is used to normalize the shapes of a given
+  weight array. 
+
+  Parameters
+  ----------
+  w: np.ndarray
+    `w` can be `None` or a scalar or a `np.ndarray` of shape
+    `(n_samples,)` or of shape `(n_samples, n_tasks)`. If `w` is a
+    sclar, it's assumed to be the same weight for all samples/tasks.
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py]: Will fix!

File: deepchem/utils/evaluate.py (hunk 6 above)
Comment (rbharath) [reply deepchem/utils/evaluate.py]: Good suggestion! Have swapped over to a new name

File: deepchem/metrics/__init__.py (hunk 2 above)
Comment (rbharath) [reply deepchem/metrics/__init__.py]: I've swapped over defaults for `n_classes` to `None`

File: deepchem/metrics/__init__.py (hunk 3 above)
Comment (rbharath) [reply deepchem/metrics/__init__.py:774]: Ah good point! Will fix the docs for this

File: deepchem/metrics/__init__.py [hunk 12]
```diff
@@ -19,11 +20,280 @@
...
+        n_tasks = 1
+        y_out = np.expand_dims(y, 1)
+      elif len(y.shape) == 3:
+        y_out = y
+      else:
+        raise ValueError(
+            "y must be an array of dimension 1, 2, or 3 for classification problems."
+        )
+    else:
+      # In this clase, y is a scalar. We assume that `y` is binary
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py:200]: Good suggestion! Will swap this to raise an error

File: deepchem/metrics/__init__.py [hunk 13]
```diff
@@ -19,11 +20,280 @@
...
+    If `mode=="classification"`, `y_out` is an array of shape `(N,
+    n_tasks, n_classes)`. If `mode=="regression"`, `y_out` is an array
+    of shape `(N, n_tasks)`.
+  """
+  if mode == "classification":
+    if n_classes is None:
+      if isinstance(y, np.ndarray):
+        # Find number of classes. Note that `y` must have values in
+        # range 0 to n_classes - 1
+        n_classes = np.amax(y) + 1
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py]: Ah good point! I'll fix this to handle logits as well

File: deepchem/metrics/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py:167]: Good suggestion! I'll aggregate the checks at the top of the function

File: deepchem/metrics/__init__.py [hunk 14]
```diff
@@ -314,132 +609,131 @@ def __init__(self,
...
-    self.mode = mode
-    # The convention used is that the first task is the metric.
-    # TODO(rbharath, joegomes): This doesn't seem like it should be hard-coded as
-    # an option in the Metric class. Instead, this should be possible to move into
-    # user-space as a custom task_averager function.
-    self.compute_energy_metric = compute_energy_metric
+        logger.info(
+            "Could not detect mode of classifier. Check your results carefully."
+        )
+      self.mode = mode
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py:741]: Ah good catch! This was incorrect. Will fix

File: deepchem/metrics/__init__.py [hunk 15]
```diff
@@ -314,132 +609,131 @@ def __init__(self,
...
-      else:
-        w_task = w[:, task]
-
-      metric_value = self.compute_singletask_metric(y_task, y_pred_task, w_task)
+      w_task = w[:, task]
+      if threshold is not None:
+        y_task = threshold_predictions(y_task, threshold=threshold)
+        y_task = to_one_hot(y_task, n_classes=n_classes)
+        y_pred_task = threshold_predictions(y_pred_task, threshold=threshold)
+        y_pred_task = to_one_hot(y_pred_task, n_classes=n_classes)
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py]: I've cleaned it up so this should be gone now

File: deepchem/metrics/__init__.py [hunk 16]
```diff
@@ -270,38 +543,60 @@ def __init__(self,
...
+      logger.warn(
+          "threshold is deprecated and will be removed in a future version of DeepChem. Set threshold in compute_metric instead"
+      )
+    if compute_energy_metric is not None:
+      self.compute_energy_metric = compute_energy_metric
+      logger.warn(
+          "compute_energy_metric is deprecated and will be removed in a future version of DeepChem."
+      )
+    else:
+      self.compute_energy_metric = False
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py:672]: Will do!

File: deepchem/metrics/__init__.py [hunk 17]
```diff
@@ -270,38 +543,60 @@ def __init__(self,
...
+        else:
+          self.name = "unknown metric"
       else:
-        self.name = self.task_averager.__name__ + "-" + self.metric.__name__
+        if hasattr(self.metric, '__name__'):
+          self.name = task_averager.__name__ + "-" + self.metric.__name__
+        else:
+          self.name = "unknown metric"
     else:
       self.name = name
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py:690]: Good catch! Will clean this up

[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/metrics/__init__.py:690]: Looking at this more closely, I think the original structure is needed. The main reason is we're casing on the argument `task_averager` (which may still be `None`) rather than `self.task_averager` which is guaranteed to not be `None`. This handling is needed to handle singletask metrics correctly.

**Comment by rbharath** [comment]
@peastman @nd-02110114 I think this PR is ready for a second round of review. I made a number of fixes to the previous code to clean up failing tests and handle edge cases. (This PR looks a lot bigger, but it actually isn't. To facilitiate debugging, I had to convert our test suites over to pytest function style instead of the old unittest since pytest can't run a single unittest at a time)

The core principle is that the `normalize_X` and `handle_classification_mode` functions handle the many different variants of tasks/shapes possible. This PR was very challenging since until this set of fixes, DeepChem didn't properly handle multitask multiclass data. I believe that I've caught all the edge cases, but this was some very tricky code so it's possible I've missed something. Feedback would be much appreciated!

**Comment by rbharath** [comment]
There's one minor failing test which I'll fix tomorrow

**Comment by nd-02110114** [comment]
@rbharath 
Maybe I will be able to check this PR tomorrow.  If you're in a hurry, it's okay to merge 🙆‍♂️

**Comment by rbharath** [comment]
@nd-02110114 Good suggestions! I think they would fit well in a subsequent PR

Tests are green so I'm going to go ahead and merge this one in
